### PR TITLE
Add MiniCPM, Deepseek V2 chat template + clean up `llama_chat_apply_template_internal`

### PR DESCRIPTION
**Pull request by ngxson**
Replaces #6236 

This PR replaces the `tmpl.find(haystack) != std::string::npos` pattern with a lambda function `tmpl_contains(haystack)`

Also add MiniCPM chat template

```
You are a helpful assistant<用户>Hello<AI>Hi there<用户>Who are you<AI>I am an assistant<用户>Another question<AI>
```

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low

### CODE REVIEW

**Comment by ngxson** [comment]
@fairydreaming I added deepseek chat template to this PR. Without chat template, it speaks somewhat english, but now it only speaks chinese:

```
> hi, who are you
我是DeepSeek Coder，一个由深度求索公司开发的智能助手。我可以帮助您回答问题和提供信息。请问有什么可以帮助您的吗？

> nice
很高兴您觉得还可以！如果您有任何问题或需要帮助的地方，请随时告诉我。

>
```

File: src/llama.cpp [hunk 1]
```diff
@@ -19598,6 +19601,35 @@ static int32_t llama_chat_apply_template_internal(
...
+            if (role == "system") {
+                ss << message->content << "\n\n";
+            } else if (role == "user") {
+                ss << "User: " << message->content << "\n\n";
+            } else if (role == "assistant") {
+                ss << "Assistant: " << message->content << eos_token;
+            }
+        }
+        if (add_ass) {
+            ss << "Assistant: ";
```
Comment (fairydreaming) [inline src/llama.cpp:19631]: I think you have extra space after "Assistant:" here, this causes the model to speak Chinese.

**Review by fairydreaming — CHANGES_REQUESTED** [review]
@ngxson I'm not sure if it's a good idea to call it deepseek-lite chat template. From what I see the following models:
- https://huggingface.co/deepseek-ai/DeepSeek-V2
- https://huggingface.co/deepseek-ai/DeepSeek-V2-Chat
- https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite
- https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite-Chat
- https://huggingface.co/deepseek-ai/DeepSeek-Coder-V2-Instruct
- https://huggingface.co/deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct

all use the same chat template in tokenizer_config.json, so it's better to call it deepseek2. DeepSeek-V2 was first to use it, so I think it's best to refer in comments to simply DeepSeek-V2 instead of DeepSeek-Coder-V2-Lite-Instruct-GGUF like you did.

File: src/llama.cpp [hunk 2]
```diff
@@ -19598,6 +19601,35 @@ static int32_t llama_chat_apply_template_internal(
...
+            std::string role(message->role);
+            if (role == "user") {
+                ss << user_tag;
+                ss << trim(message->content);
+                ss << "<AI>";
+            } else {
+                ss << trim(message->content);
+            }
+        }
+    } else if (tmpl == "deepseek-lite" || tmpl_contains("'Assistant: ' + message['content'] + eos_token")) {
```
Comment (fairydreaming) [inline src/llama.cpp:19617]: Change "deepseek-lite" to "deepseek2"

File: tests/test-chat-template.cpp [hunk 3]
```diff
@@ -94,6 +98,10 @@ int main(void) {
         "<|system|>\nYou are a helpful assistant<|end|>\n<|user|>\nHello<|end|>\n<|assistant|>\nHi there<|end|>\n<|user|>\nWho are you<|end|>\n<|assistant|>\n   I am an assistant   <|end|>\n<|user|>\nAnother question<|end|>\n<|assistant|>\n",
         //Phi-3-vision
         "<|system|>\nYou are a helpful assistant<|end|>\n<|user|>\nHello<|end|>\n<|assistant|>\nHi there<|end|>\n<|user|>\nWho are you<|end|>\n<|assistant|>\n   I am an assistant   <|end|>\n<|user|>\nAnother question<|end|>\n<|assistant|>\n",
+        // MiniCPM-3B-OpenHermes-2.5-v2-GGUF
+        u8"You are a helpful assistant<用户>Hello<AI>Hi there<用户>Who are you<AI>I am an assistant<用户>Another question<AI>",
+        // DeepSeek-Coder-V2-Lite-Instruct-GGUF
+        u8"You are a helpful assistant\n\nUser: Hello\n\nAssistant: Hi there<｜end▁of▁sentence｜>User: Who are you\n\nAssistant:    I am an assistant   <｜end▁of▁sentence｜>User: Another question\n\nAssistant: ",
     };
```
Comment (fairydreaming) [inline tests/test-chat-template.cpp:105]: Here is also an unnecessary space after "Assistant:" at the end.

File: tests/test-chat-template.cpp [hunk 4]
```diff
@@ -94,6 +98,10 @@ int main(void) {
         "<|system|>\nYou are a helpful assistant<|end|>\n<|user|>\nHello<|end|>\n<|assistant|>\nHi there<|end|>\n<|user|>\nWho are you<|end|>\n<|assistant|>\n   I am an assistant   <|end|>\n<|user|>\nAnother question<|end|>\n<|assistant|>\n",
         //Phi-3-vision
         "<|system|>\nYou are a helpful assistant<|end|>\n<|user|>\nHello<|end|>\n<|assistant|>\nHi there<|end|>\n<|user|>\nWho are you<|end|>\n<|assistant|>\n   I am an assistant   <|end|>\n<|user|>\nAnother question<|end|>\n<|assistant|>\n",
+        // MiniCPM-3B-OpenHermes-2.5-v2-GGUF
+        u8"You are a helpful assistant<用户>Hello<AI>Hi there<用户>Who are you<AI>I am an assistant<用户>Another question<AI>",
+        // DeepSeek-Coder-V2-Lite-Instruct-GGUF
```
Comment (fairydreaming) [inline tests/test-chat-template.cpp:103]: Replace with DeepSeek-V2

File: tests/test-chat-template.cpp [hunk 5]
```diff
@@ -57,7 +57,11 @@ int main(void) {
         //Phi-3-medium
         "{% for message in messages %}{% if (message['role'] == 'user') %}{{'<|user|>' + '\n' + message['content'] + '<|end|>' + '\n' + '<|assistant|>' + '\n'}}{% elif (message['role'] == 'assistant') %}{{message['content'] + '<|end|>' + '\n'}}{% endif %}{% endfor %}",
         //Phi-3-vision
-        "{% for message in messages %}{{'<|' + message['role'] + '|>' + '\n' + message['content'] + '<|end|>\n' }}{% endfor %}{% if add_generation_prompt and messages[-1]['role'] != 'assistant' %}{{- '<|assistant|>\n' -}}{% endif %}"
+        "{% for message in messages %}{{'<|' + message['role'] + '|>' + '\n' + message['content'] + '<|end|>\n' }}{% endfor %}{% if add_generation_prompt and messages[-1]['role'] != 'assistant' %}{{- '<|assistant|>\n' -}}{% endif %}",
+        // MiniCPM-3B-OpenHermes-2.5-v2-GGUF
+        u8"{% for message in messages %}{% if message['role'] == 'user' %}{{'<用户>' + message['content'].strip() + '<AI>'}}{% else %}{{message['content'].strip()}}{% endif %}{% endfor %}",
+        // DeepSeek-Coder-V2-Lite-Instruct-GGUF
```
Comment (fairydreaming) [inline tests/test-chat-template.cpp:63]: Replace with DeepSeek-V2

File: src/llama.cpp [hunk 6]
```diff
@@ -19598,6 +19601,35 @@ static int32_t llama_chat_apply_template_internal(
...
+        // DeepSeek-Coder-V2-Lite-Instruct-GGUF
```
Comment (fairydreaming) [inline src/llama.cpp:19618]: Replace with DeepSeek-V2

File: src/llama.cpp (hunk 1 above)
Comment (ngxson) [reply src/llama.cpp]: Extra space removed in https://github.com/ggerganov/llama.cpp/pull/8172/commits/736c4948c0263b7ef5b6e50fc1f3ec2c5abb4df2

File: src/llama.cpp (hunk 2 above)
Comment (ngxson) [reply src/llama.cpp]: Changed in https://github.com/ggerganov/llama.cpp/pull/8172/commits/736c4948c0263b7ef5b6e50fc1f3ec2c5abb4df2

**Comment by ngxson** [comment]
@fairydreaming Thanks for the info. I was confused between deepseek (v1), deepseek-V2 and lite-V2. With the extra space removed, it's now speaking correct language:

```
> who are you
 I am DeepSeek Coder, an intelligent assistant developed by DeepSeek, a company in China.

> how do you say "happy new year" in chinese
 In Chinese, "happy new year" can be expressed in several ways. Here are some options:

1. 新年快乐 (Xīnnián kuàilè) - This phrase literally means "new year happy." It's the most common way to say "happy new year" in Chinese.
```

File: tests/test-chat-template.cpp (hunk 5 above)
Comment (fairydreaming) [reply tests/test-chat-template.cpp:63]: I think you forgot to change this one.

**Review by fairydreaming — APPROVED** [review]
There's still one "DeepSeek-Coder-V2-Lite-Instruct-GGUF" in the comments, correct it and should be good to go.

**Pull request merged**